### ISSUE TITLE
[bitnami/discourse] Add existingSecret sections to chart config

### ISSUE THREAD

**Issue by LucaPrete** [bitnami/charts#3035]
**Which chart**:
discourse

**Is your feature request related to a problem? Please describe.**
The chart does not allow to use existing secrets and managing db dependencies (i.e. create new DB vs use existing DBs) is quite complex. Many values for DBs are duplicated in the values.yaml file.

**Describe the solution you'd like**
I'd like to remove the externalDatabase / externalRedis sections. I'll absorb the params in the postgresql and redis sections.
I'll also introduce new values that will allow to use existing secrets, both for discourse, postgresql and redis passwords.

**Comment by LucaPrete** [comment]
Going to send a first PR...

**Comment by dani8art** [comment]
Hi @LucaPrete thanks for your feedback we will be happy to handle the PR, let's move the conversation there!

**Comment by LucaPrete** [comment]
Daniel, thanks!

The PR is ready for review :)

-Luca

Il giorno mar 7 lug 2020 alle 17:24 Daniel Arteaga <notifications@github.com>
ha scritto:

**Comment by ieugen** [comment]
I've also encountered these issues. 
Any news on the PR? 
I'm figuring out how to use the chart now and it's a bit frustrating :) .

**Comment by ieugen** [comment]
I found the usage of `externalDatabase.postgresqlPostgresPassword` not good.
I'm trying to expose an external PostgresSQL running on Debian and they don't set passwords for `postgres` user. 
You use `sudo -u postgresq psql` to operate. 

Is this password necessary when using an external database? 
I believe the chart should allow for a super user + password and not expect postgres.

**Comment by ieugen** [comment]
I would also like to give some feedback on the ingress and certificate issuance.
There are assumptions that a load ballancer is used.
Also I didn't figure out how to issue certificates with cert-manager which IMO is quite common.

**Comment by ieugen** [comment]
helm uninstall does not seem to clean the services when using an external database. 
p.s. I had a typo in my host name and that is why the certificates did not work.

**Comment by ieugen** [comment]
For database initialization I would use the "Initialization scripts" https://hub.docker.com/_/postgres  .

The script can be a shell one and it will run with super user + it can use env variables (for password). 

The commands I used are for discourse are (I initialized it manually for now) : 
```
   # Connect to db
    sudo -u postgresql
   # create suff 
    create user discourse_co with password '<password>';
    create database discourse_co with owner discourse_co;
    \c discourse_co;
    create extension hstore;
    create extension pg_trgm;
```

**Comment by ieugen** [comment]
With my setup, the admin is not created and I don't know why at the moment.
So I can't access the instance once I've set it up.
Also there seems to be no SMPT config - so no validation emails are getting out. 
Hence, I am locked out. 

I'm deploying on k3s, bare metal. 

````
global:
  storageClass: local-path

image:
  tag: 2.4.4-debian-10-r45

service:
  type: ClusterIP

persistence:
  enabled: true
  storageClass: local-path

discourse:
  host: platforma.REDACTED.com
  siteName: REDACTED
  username: discourse_admin
  email: eugen@example.com

postgresql:
  enabled: false

externalDatabase:
  host: discourse-database
  user: discourse_co
  password: "my-secret-password"
  database: discourse_co
  postgresqlPostgresPassword: "postgres"
  port: 5432

ingress:
  enabled: true
  certManager: true
  hostname: platforma.xxxxx.com
  tls: true
  annotations:
    cert-manager.io/cluster-issuer: cloudflare-cluster-issuer
    traefik.ingress.kubernetes.io/redirect-entry-point: https
    acme.cert-manager.io/http01-ingress-class: traefik
````

**Comment by dani8art** [comment]
Hi @ieugen, thank you so much for your great feedback!!

I think there is so many things, let's go step by step

> I found the usage of externalDatabase.postgresqlPostgresPassword not good.
> I'm trying to expose an external PostgresSQL running on Debian and they don't set passwords for postgres user.

Yes your are right, they shouldn't be mandatory when using `externalDatabase` we can fix it as follow and you can find it in this PR (#3125).

**values.yaml**

```yaml
  ## Database password
  ##
  password: ""
  ## Root "postgres" password (used in the installation stage)
  ##
  postgresqlPostgresPassword: ""
```

> There are assumptions that a load ballancer is used.
> Also I didn't figure out how to issue certificates with cert-manager which IMO is quite common

Could you open an issue and tell us with more details what does this mean?

> helm uninstall does not seem to clean the services when using an external database.

I'm not sure about that, I've just tried to reproduce but helm always delete my services.

> For database initialization I would use the "Initialization scripts"

In fact, you could use it if you use `postgresql.enabled=true` as it is part of `bitnami/posgresql` docker image too. Please see more https://github.com/bitnami/charts/tree/master/bitnami/postgresql#initialize-a-fresh-instance

> With my setup, the admin is not created and I don't know why at the moment.

Could you open an issue and tell us with more details what does this mean?

**Comment by ieugen** [comment]
Hi @dani8art , 

Thanks for the feedback. I took a break yesterday. 
I can open some issues but right now I don't know if they are bugs with the chart or just me not knowing how to properly use the chart (documentation issue).

I fixed the issue with certificate. 
I figured out I need to pass in SMTP configuration and found the values nicely documented here https://github.com/bitnami/bitnami-docker-discourse#environment-variables . 

However I can't get them to the container. 

With this setup I get an error: 
```
discourse:
  host: platforma.YYY.com
  siteName: XXXXX
  username: admin
  email: me@example.com
  extraEnvVars:
    - name: SMTP_HOST
      value: "smtp.example.com"
    - name: SMTP_PORT
      value: "587"

helm upgrade ....

Error: UPGRADE FAILED: YAML parse error on discourse/templates/deployment.yaml: error converting YAML to JSON: yaml: line 57: mapping values are not allowed in this context
helm.go:75: [debug] error converting YAML to JSON: yaml: line 57: mapping values are not allowed in this context
```

Some general feedback, not related to the above issue:

I've read a bit on how to deploy discourse and they have a setup that is not very suited for deploying in K8s. 
They package the app and all it's dependencies which makes foe easier management as a bundle however, in k8s with scaling and such you might get into conflicts since postgresqla and redis are not scalable just like that (tm).

It's my believe that the chart should always use external database and external redis.

Having them internally is bringing a world of trouble and it's impossible to reason about. 
What happens to postgresql or dedis (storage) when the app scales up / down ? 
What about on upgrades when the new version and the old version are running side by side until the k8s probes determine it's time to make the switch? 

To me it' either switch the app to a StatefullSet or just drop all internal postgresql / redis stuff and always use external.
This way the chart is Waaaay more simple to reason about and providing an existing postgresql and redis is left to the user.
For that task there are specialized charts that deal with the above issues (hopefully).
Also the chart will be much more simple to maintain.

**Comment by ieugen** [comment]
I downloaded the chart locally with `helm pull bitnami/discourse` and added the SMTP env vars manually.
The emails still did not reach me - sendmail is missing :( . 

`kubectl  logs -f discourse-co-76465d5c7b-pz9sk sidekiq`
````
./scripts/ctl.sh : discourse_sidekiq started

sh: 1: /usr/sbin/sendmail: not found
sh: 1: /usr/sbin/sendmail: not found
````

**Comment by dani8art** [comment]
Hi @ieugen, 

thanks for continue given us feedback!

> Error: UPGRADE FAILED: YAML parse error on discourse/templates/deployment.yaml: error converting YAML to JSON: yaml: line 57: mapping values are not allowed in this context

Should be fixed by #3137

> sendmail is missing

I wasn't able to reproduce it, please could you try to open a new issue so we can go deeper on it?
